### PR TITLE
late initialization and weight redistribution

### PR DESCRIPTION
**Pull request by kalantar**

### CODE REVIEW

File: controllers/metrics.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+func (r *ExperimentReconciler) readMetricsInternal(ctx context.Context, instance *v2alpha1.Experiment) bool {
+	log := util.Logger(ctx)
+	log.Info("readMetricsInternal() called")
+	defer log.Info("readMetricsInternal() completed")
+
+	criteria := instance.Spec.Criteria
+	if criteria == nil {
+		r.markMetricsSynced(ctx, instance, "No criteria specified")
+		r.markSpecUpdated()
```
Comment (sriumcp) [inline controllers/metrics.go:86]: markSpecUpdated() needs to change to markStatusUpdated()

Comment (sriumcp) [inline controllers/metrics.go:86]: Is it already updated? Or is there a pending update?

File: controllers/recorder.go [hunk 2]
```diff
@@ -31,41 +37,65 @@ import (
...
-func (r *ExperimentReconciler) markIterationUpdate(context context.Context, instance *v2alpha1.Experiment,
+func (r *ExperimentReconciler) markIterationUpdate(ctx context.Context, instance *v2alpha1.Experiment,
 	messageFormat string, messageA ...interface{}) {
 	if updated, reason := instance.Status.MarkIterationUpdate(messageFormat, messageA...); updated {
-		util.Logger(context).Info(reason + ", " + fmt.Sprintf(messageFormat, messageA...))
+		util.Logger(ctx).Info(reason + ", " + fmt.Sprintf(messageFormat, messageA...))
 		// record event
 		// send notifications
 		r.markStatusUpdated()
 		r.markSpecUpdated()
```
Comment (sriumcp) [inline controllers/recorder.go:91]: markSpecUpdated() doesn't belong here.

**Review by sriumcp — CHANGES_REQUESTED** [review]

File: controllers/experiment_controller.go [hunk 3]
```diff
@@ -132,28 +125,39 @@ func (r *ExperimentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
-	// 	// change := instance.InitializeSpec(ctx)
-	// 	// if change {
-	// 	// 	if err := r.Update(ctx, instance); err != nil {
-	// 	// 		log.Error(err, "Failed to update spec when initializing experiment")
-	// 	// 		return ctrl.Result{}, err
-	// 	// 	}
-	// 	// }
-	// 	// log.Info("Updated spec")
-	// }
+	// LATE INITIALIZATION of spec
```
[the remark this answers is not in the document]
Comment (kalantar) [reply controllers/experiment_controller.go:128]: I've made some (minor) changes in names and approach. Let's review again. If we decide we should still do this, I would like to defer it.

File: controllers/experiment_controller.go [hunk 4]
```diff
@@ -132,28 +125,39 @@ func (r *ExperimentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
+	change := instance.InitializeSpec()
+	log.Info("initialized spec", "change", change, "updatedStatus", r.StatusUpdated)
+	log.Info("reading metrics", "conditions", instance.Status.Conditions)
+	log.Info("reading metrics", "condition", instance.Status.GetCondition(v2alpha1.ExperimentConditionMetricsSynced))
+	change = r.ReadMetrics(ctx, instance) || change
```
[the remark this answers is not in the document]
Comment (kalantar) [reply controllers/experiment_controller.go]: I would like to defer this. I think what we have will get us to a fuller demonstration and then we can fix. In the meantime, we should introduce this concept for any new errors we want to capture in a similar way.

File: controllers/experiment_controller.go [hunk 5]
```diff
@@ -132,28 +125,39 @@ func (r *ExperimentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
+	log.Info("read metrics", "change", change, "updatedStatus", r.StatusUpdated)
+	log.Info("read metrics", "conditions", instance.Status.Conditions)
+	log.Info("read metrics", "condition", instance.Status.GetCondition(v2alpha1.ExperimentConditionMetricsSynced))
+	if r.StatusUpdated {
```
[the remark this answers is not in the document]
Comment (kalantar) [reply controllers/experiment_controller.go]: done

File: controllers/experiment_controller.go [hunk 6]
```diff
@@ -210,54 +214,48 @@ func (r *ExperimentReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	r.updateIfNeeded(ctx, instance)
 
 	// trigger next experiment
 
 	return ctrl.Result{}, nil
 }
 
-func (r *ExperimentReconciler) finishExperiment(context context.Context, instance *v2alpha1.Experiment) (ctrl.Result, error) {
-	log := util.Logger(context)
+func (r *ExperimentReconciler) finishExperiment(ctx context.Context, instance *v2alpha1.Experiment) (ctrl.Result, error) {
```
[the remark this answers is not in the document]
Comment (kalantar) [reply controllers/experiment_controller.go:229]: documented in the code

File: controllers/metrics.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+		// could not read metric; skip it
+		return err
+	}
+
+	metricMap[key] = metric
+	return nil
+}
+
+// ReadMetrics finds and reads all the metrics into the spec
+func (r *ExperimentReconciler) ReadMetrics(ctx context.Context, instance *v2alpha1.Experiment) bool {
```
[the remark this answers is not in the document]
Comment (kalantar) [reply controllers/metrics.go:71]: done

File: controllers/metrics.go (hunk 1 above)
Comment (kalantar) [reply controllers/metrics.go]: done

Comment (kalantar) [reply controllers/metrics.go]: renamed and eliminated the use of a method so hopefully clearer.

File: controllers/recorder.go (hunk 2 above)
Comment (kalantar) [reply controllers/recorder.go]: done

**Comment by kalantar** [comment]
I've pushed changes to address of the comments in a new commit.

**Review by sriumcp — APPROVED** [review]
LGTM for now.

**Pull request merged**